lms/security.py

```diff
@@ -87,13 +87,25 @@ def forget(self, request):
     @lru_cache(maxsize=1)
     def get_policy(request: Request):
         """Pick the right policy based the request's path."""
-        # pylint:disable=too-many-return-statements
+        # pylint:disable=too-many-return-statements,too-complex
 
         path = request.path
 
         if path.startswith("/admin") or path.startswith("/googleauth"):
+            # Routes that require the Google auth policy
             return LMSGoogleSecurityPolicy()
 
+        if path.startswith("/dashboard/assignment/") or path.startswith(
+            "/api/assignment/"
+        ):
+            # For certain routes we only use the google policy in case it resulted
+            # non empty identity.
+            # This is useful for routes that can be used by admin pages users on top of
+            # other type of access
+            policy = LMSGoogleSecurityPolicy()
+            if policy.identity(request):
+                return policy
```

Comment:
I have noticed this can lead to a couple "interesting" scenarios.

1. I have logged in the admin as STAFF. Once there I open the dashboard for one assignment.

    Then I go to the same assignment in the corresponding LMS (Canvas in this case). I delete my local `authorization` cookie, and navigate to the dashboard.

    Once in the dashboard, if I delete the `authorization` cookie again, it keeps recreating it when reloading the page, as long as I have a valid `session` cookie. Only if I also delete the `session` cookie, then I see an authorization error that can only be fixed by re-opening the dashboard from the assignment to get a new `authorization` cookie.

2. Another even weirder combination. If I open the dashboard from the admin, let the `authorization` cookie to be created, and then delete the `session` cookie but not the `authorization` one, then I get a 500 error when reloading the page, with this trace:

    ```
    web (stderr)         | Traceback (most recent call last):
    web (stderr)         |   File "/home/alejandro/Development/hypothesis/lms/.tox/dev/lib/python3.11/site-packages/pyramid_exclog/__init__.py", line 157, in exclog_tween
    web (stderr)         |     response = handler(request)
    web (stderr)         |                ^^^^^^^^^^^^^^^^
    web (stderr)         |   File "/home/alejandro/Development/hypothesis/lms/.tox/dev/lib/python3.11/site-packages/pyramid/router.py", line 143, in handle_request
    web (stderr)         |     response = _call_view(
    web (stderr)         |                ^^^^^^^^^^^
    web (stderr)         |   File "/home/alejandro/Development/hypothesis/lms/.tox/dev/lib/python3.11/site-packages/sentry_sdk/integrations/pyramid.py", line 91, in sentry_patched_call_view
    web (stderr)         |     return old_call_view(registry, request, *args, **kwargs)
    web (stderr)         |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    web (stderr)         |   File "/home/alejandro/Development/hypothesis/lms/.tox/dev/lib/python3.11/site-packages/pyramid/view.py", line 674, in _call_view
    web (stderr)         |     response = view_callable(context, request)
    web (stderr)         |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    web (stderr)         |   File "/home/alejandro/Development/hypothesis/lms/.tox/dev/lib/python3.11/site-packages/pyramid/config/views.py", line 170, in attr_view
    web (stderr)         |     return view(context, request)
    web (stderr)         |            ^^^^^^^^^^^^^^^^^^^^^^
    web (stderr)         |   File "/home/alejandro/Development/hypothesis/lms/.tox/dev/lib/python3.11/site-packages/pyramid/config/views.py", line 196, in predicate_wrapper
    web (stderr)         |     return view(context, request)
    web (stderr)         |            ^^^^^^^^^^^^^^^^^^^^^^
    web (stderr)         |   File "/home/alejandro/Development/hypothesis/lms/.tox/dev/lib/python3.11/site-packages/pyramid/viewderivers.py", line 326, in secured_view
    web (stderr)         |     raise HTTPForbidden(msg, result=result)
    web (stderr)         | pyramid.httpexceptions.HTTPForbidden: Unauthorized: DashboardViews failed permission check
    ```

These are of course edge cases that should not happen to real users, but maybe it's worth knowing about these and documenting them.

Reply:
I don't think that `1.` needs to be handled differently, deleting the auth token results in a 403 and to gain entry again you need to reauthenticate.


2. It's arguably a token, the issue is that the LTIUser cookie is being set when there's no LTIUser to serialize. I'll handle that on a separate PR

Reply:
> I don't think that `1.` needs to be handled differently, deleting the auth token results in a 403 and to gain entry again you need to reauthenticate.

Yes, I don't think it is working "incorrectly", it's more that it could lead to false positive cases in which you go to a dashboard thinking you are logged-in as instructor, everything works as you were expecting, but then it turns out you were logged in because you had a `session` cookie from before, and for teachers it didn't actually work.

Anyway, it's pretty much an edge case and I don't think we can really handle it.

> `2` It's arguably a bug, the issue is that the LTIUser cookie is being set when there's no LTIUser to serialize. I'll handle that on a separate PR

👍🏼